In [2]:
import pandas as pd
from sqlalchemy import create_engine

In [4]:
# establish a database connection
engine = create_engine("postgresql+psycopg2://postgres:postgres@localhost:5432/lahman_baseball")
# Replace `<lahman_baseball_database_name>` with the actual name of your lahman baseball database as it appears in pgadmin

# use the connection to run a query using pandas!
df = pd.read_sql("SELECT * FROM batting;", con=engine)
df.head()

playerid  yearid  stint teamid lgid   g   ab   r   h  h2b  ...   rbi   sb  \
0  abercda01    1871      1    TRO   NA   1    4   0   0    0  ...   0.0  0.0   
1   addybo01    1871      1    RC1   NA  25  118  30  32    6  ...  13.0  8.0   
2  allisar01    1871      1    CL1   NA  29  137  28  40    4  ...  19.0  3.0   
3  allisdo01    1871      1    WS3   NA  27  133  28  44   10  ...  27.0  1.0   
4  ansonca01    1871      1    RC1   NA  25  120  29  39   11  ...  16.0  6.0   

    cs  bb   so  ibb  hbp  sh  sf  gidp  
0  0.0   0  0.0  NaN  NaN NaN NaN   NaN  
1  1.0   4  0.0  NaN  NaN NaN NaN   NaN  
2  1.0   2  5.0  NaN  NaN NaN NaN   NaN  
3  1.0   0  2.0  NaN  NaN NaN NaN   NaN  
4  2.0   2  1.0  NaN  NaN NaN NaN   NaN  

[5 rows x 22 columns]

## Problem 1: Vanderbilt

a. Find all players in the database who played at Vanderbilt University.

b. Create a list showing each player’s first and last names as well as the total salary they earned in the major leagues.

c. Sort this list in descending order by the total salary earned.

d. Which Vanderbilt player earned the most money in the majors?

## Import

In [5]:
people = pd.read_sql("SELECT * FROM people", con=engine)
salaries = pd.read_sql("SELECT * FROM salaries", con = engine)
collegeplaying = pd.read_sql("SELECT * FROM collegeplaying", con = engine)
schools = pd.read_sql("SELECT * FROM schools", con = engine)

### Part (a)

In [22]:
#filter schools for vanderbilt
vandy = schools[schools['schoolname'] == "Vanderbilt University"]

#merge vandy to collegeplaying to make vandycp
vandycp = pd.merge(vandy, collegeplaying, how = 'inner', on = 'schoolid')
vandycp.drop('yearid', axis = 'columns', inplace = True)
vandycp = vandycp.drop_duplicates()

#merge vandycp to people to make vandypeople
vandypeople = pd.merge(vandycp, people, how = 'inner', on = 'playerid')

#list namefirst and namelast
print(vandypeople[['namefirst', 'namelast']])

   namefirst    namelast
0      Pedro     Alvarez
1       Mike      Baxter
2       Nick  Christiani
3     Wilson     Collins
4       Joey        Cora
5       Slim       Embry
6       Ryan    Flaherty
7      Sonny        Gray
8     Harvey    Hendrick
9       Matt        Kata
10    Jensen       Lewis
11    Scotti     Madison
12      Mike       Minor
13   Scrappy       Moore
14       Mal        Moss
15      Josh        Paul
16     David       Price
17      Mark       Prior
18    Antoan  Richardson
19     Scott   Sanderson
20       Rip      Sewell
21    Jeremy      Sowers
22      Mike      Willis
23      Josh        Zeid


### Part (b)

In [35]:
#merge vandypeople (just three relevant columns) with salaries to get vandy_sals
vandy_sals = pd.merge(vandypeople[['playerid','namefirst', 'namelast']], 
                      salaries, 
                      how = 'left', 
                      on = 'playerid')
vandy_sal_sum = vandy_sals.groupby(['playerid', 'namelast','namefirst'], as_index = False).sum()
vandy_sal_sum.drop(['playerid','yearid'], axis = 'columns', inplace = True)
vandy_sal_sum

namelast namefirst      salary
0      Alvarez     Pedro  20681704.0
1       Baxter      Mike   2094418.0
2   Christiani      Nick    500000.0
3      Collins    Wilson         0.0
4         Cora      Joey   5622500.0
5        Embry      Slim         0.0
6     Flaherty      Ryan   4061000.0
7         Gray     Sonny   1542500.0
8     Hendrick    Harvey         0.0
9         Kata      Matt   1060000.0
10       Lewis    Jensen   1234000.0
11     Madison    Scotti    135000.0
12       Minor      Mike   6837500.0
13       Moore   Scrappy         0.0
14        Moss       Mal         0.0
15        Paul      Josh   2640000.0
16       Price     David  81851296.0
17       Prior      Mark  12800000.0
18  Richardson    Antoan         0.0
19   Sanderson     Scott  10750000.0
20      Sewell       Rip         0.0
21      Sowers    Jeremy    384800.0
22      Willis      Mike         0.0
23        Zeid      Josh         0.0

### Part (c)

In [39]:
vandy_sal_sum.sort_values('salary', ascending = False, ignore_index = True)

namelast namefirst      salary
0        Price     David  81851296.0
1      Alvarez     Pedro  20681704.0
2        Prior      Mark  12800000.0
3    Sanderson     Scott  10750000.0
4        Minor      Mike   6837500.0
5         Cora      Joey   5622500.0
6     Flaherty      Ryan   4061000.0
7         Paul      Josh   2640000.0
8       Baxter      Mike   2094418.0
9         Gray     Sonny   1542500.0
10       Lewis    Jensen   1234000.0
11        Kata      Matt   1060000.0
12  Christiani      Nick    500000.0
13      Sowers    Jeremy    384800.0
14     Madison    Scotti    135000.0
15       Embry      Slim         0.0
16      Willis      Mike         0.0
17      Sewell       Rip         0.0
18        Moss       Mal         0.0
19  Richardson    Antoan         0.0
20     Collins    Wilson         0.0
21       Moore   Scrappy         0.0
22    Hendrick    Harvey         0.0
23        Zeid      Josh         0.0

### Part (d)

In [40]:
#David Price, with a wopping $81,851,296

## Problem 2: Fielding Groups

Using the fielding table, group players into three groups based on their position: label players with position OF as "Outfield", those with position "SS", "1B", "2B", and "3B" as "Infield", and those with position "P" or "C" as "Battery". Determine the number of putouts made by each of these three groups in 2016.

In [41]:
fielding = pd.read_sql("SELECT * FROM fielding", con = engine)

         playerid  yearid  stint teamid lgid pos    g     gs  innouts   po  \
0       abercda01    1871      1    TRO   NA  SS    1    NaN      NaN    1   
1        addybo01    1871      1    RC1   NA  2B   22    NaN      NaN   67   
2        addybo01    1871      1    RC1   NA  SS    3    NaN      NaN    8   
3       allisar01    1871      1    CL1   NA  2B    2    NaN      NaN    1   
4       allisar01    1871      1    CL1   NA  OF   29    NaN      NaN   51   
...           ...     ...    ...    ...  ...  ..  ...    ...      ...  ...   
136810  zobribe01    2016      1    CHN   NL  2B  119  113.0   2929.0  177   
136811  zobribe01    2016      1    CHN   NL  OF   46   29.0    859.0   43   
136812  zobribe01    2016      1    CHN   NL  SS    1    0.0      6.0    0   
136813  zuninmi01    2016      1    SEA   AL   C   52   48.0   1331.0  400   
136814   zychto01    2016      1    SEA   AL   P   12    0.0     41.0    0   

            a     e    dp   pb  wp    sb   cs  zr  
0         3

In [53]:
f_pos_gp_ls = []
for i in fielding['pos']:
    if i in ["SS","1B","2B","3B"]:
        pos_gp = "infield"
    elif i in ["P", "C"]:
        pos_gp = "Battery"
    else:
        pos_gp = "Outfield"
    f_pos_gp_ls.append(pos_gp)
    
fielding["pos_group"] = f_pos_gp_ls

fielding2016 = fielding[fielding['yearid'] == 2016]

fielding2016[['po', 'pos_group']].groupby('pos_group', as_index = False).sum()

pos_group     po
0   Battery  41424
1  Outfield  29560
2   infield  58934

## Problem 3: World Series vs Wins

a. From 1970 – 2016, what is the largest number of wins for a team that did not win the world series?
b. What is the smallest number of wins for a team that did win the world series?
c. Doing this will probably result in an unusually small number of wins for a world series champion – determine why this is the case.
d. Then redo your query, excluding the problem year.
e. How often from 1970 – 2016 was it the case that a team with the most wins also won the world series?
f. What percentage of the time?

### Part (a)

In [54]:
#import teams as df
teams = pd.read_sql("SELECT * FROM teams", con = engine)
print(teams)

      yearid lgid teamid franchid divid  rank    g  ghome   w   l  ...     dp  \
0       1871   NA    BS1      BNA  None     3   31    NaN  20  10  ...    NaN   
1       1871   NA    CH1      CNA  None     2   28    NaN  19   9  ...    NaN   
2       1871   NA    CL1      CFC  None     8   29    NaN  10  19  ...    NaN   
3       1871   NA    FW1      KEK  None     7   19    NaN   7  12  ...    NaN   
4       1871   NA    NY2      NNA  None     5   33    NaN  16  17  ...    NaN   
...      ...  ...    ...      ...   ...   ...  ...    ...  ..  ..  ...    ...   
2830    2016   NL    SLN      STL     C     2  162   81.0  86  76  ...  169.0   
2831    2016   AL    TBA      TBD     E     5  162   81.0  68  94  ...  129.0   
2832    2016   AL    TEX      TEX     W     1  162   81.0  95  67  ...  190.0   
2833    2016   AL    TOR      TOR     E     2  162   81.0  89  73  ...  144.0   
2834    2016   NL    WAS      WSN     E     1  162   81.0  95  67  ...  142.0   

         fp                

In [65]:
#limit teams to relevant columns: teams2
teams2 = teams[['yearid','teamid', 'name', 'g', 'w', 'wswin']]

#limit teams2 to relevant years: team_recent

teams_recent = teams2[teams2['yearid'] >= 1970]

#create two separate dataframes, one for teams that won the world series
#and one for teams that did not win world series

teams_no_wsw = teams_recent[teams_recent['wswin'] == 'N']
teams_wsw = teams_recent[teams_recent['wswin'] == 'Y']

#find world series no-win row with maximum wins
teams_no_wsw[teams_no_wsw['w'] == teams_no_wsw['w'].max()]

yearid teamid              name    g    w wswin
2379    2001    SEA  Seattle Mariners  162  116     N

### Part (b)

In [66]:
#find world series win row with minimum wins
teams_wsw[teams_wsw['w'] == teams_wsw['w'].min()]

yearid teamid                 name    g   w wswin
1824    1981    LAN  Los Angeles Dodgers  110  63     Y

### Part (c)

There was a lockout in 1981, which reduced the number games in the season (only 110).

### Part (d)

In [67]:
#teams_wsw[(teams_wsw['yearid'] != 1981) & (teams_wsw['w'] == teams_wsw['w'].min())]
teams_wsw2 = teams_wsw[teams_wsw['yearid'] != 1981]
teams_wsw2[teams_wsw2['w'] == teams_wsw2['w'].min()]

yearid teamid                 name    g   w wswin
2530    2006    SLN  St. Louis Cardinals  161  83     Y

### Part (e)

In [93]:
#List of teams with most wins in season and who won the world series that season
teams_recent[(teams_recent.groupby('yearid')['w'].transform('max') == teams_recent['w'])
            &
            (teams_recent['wswin'] == 'Y')]

yearid teamid               name    g    w wswin
1542    1970    BAL  Baltimore Orioles  162  108     Y
1667    1975    CIN    Cincinnati Reds  162  108     Y
1691    1976    CIN    Cincinnati Reds  162  102     Y
1750    1978    NYA   New York Yankees  163  100     Y
1899    1984    DET     Detroit Tigers  162  104     Y
1959    1986    NYN      New York Mets  162  108     Y
2038    1989    OAK  Oakland Athletics  162   99     Y
2283    1998    NYA   New York Yankees  162  114     Y
2538    2007    BOS     Boston Red Sox  162   96     Y
2612    2009    NYA   New York Yankees  162  103     Y
2718    2013    BOS     Boston Red Sox  162   97     Y
2810    2016    CHN       Chicago Cubs  162  103     Y

In [97]:
#count of teams in that list
teams_recent[(teams_recent.groupby('yearid')['w'].transform('max') == teams_recent['w'])
            &
            (teams_recent['wswin'] == 'Y')].count()

yearid    12
teamid    12
name      12
g         12
w         12
wswin     12
dtype: int64

### Part (f)

In [100]:
#percentage of most-win teams to win ws from 1970 - 2016:

#number of teams with most wins in a season (sometimes more than one per year)
topw = teams_recent[teams_recent.groupby('yearid')['w'].transform('max') == teams_recent['w']]

#number of teams with most wins who won world series
topw_ws = teams_recent[(teams_recent.groupby('yearid')['w'].transform('max') == teams_recent['w'])
            &
            (teams_recent['wswin'] == 'Y')]

#divide, multiply by 100, append a % sign
print(topw_ws['teamid'].count() * 100/ topw['teamid'].count(), '%')

22.641509433962263 %


## Problem 4: Manager of the Year

Which managers have won the TSN Manager of the Year award in both the National League (NL) and the American League (AL)? Give their full name and the teams that they were managing when they won the award.

In [101]:
awardsmgrs = pd.read_sql('SELECT * FROM awardsmanagers', con = engine)
mgrs = pd.read_sql('SELECT * FROM managers', con = engine)

In [114]:
TSN = awardsmgrs[awardsmgrs['awardid'] == 'TSN Manager of the Year']

#filter out leaguewide ('ML') results (sorry Bobby Cox)
TSN_noML = TSN[TSN['lgid'] != "ML"]

#keep only managers whose distinct lgid count is 2
TSN_ALNL = TSN_noML[(TSN_noML.groupby(['playerid'])['lgid'].transform('nunique') == 2)]

#merge with people to get namefirst and namelast
TSN_ALNLp = TSN_ALNL.merge(people, how = 'inner', on = 'playerid')[['playerid', 'awardid', 'yearid', 'lgid', 'namefirst', 'namelast']]

#merge with mgrs to get teamid by yearid
TSN_ALNL_det = TSN_ALNLp.merge(mgrs, how = 'inner', on = ['playerid', 'yearid'])

#merge with teams to get team names
TSN_ALNL_complete = TSN_ALNL_det.merge(teams, how = 'inner', on = ['yearid', 'teamid'])[['playerid', 'awardid', 'yearid', 'lgid', 'namefirst', 'namelast', 'name']]

print(TSN_ALNL_complete)

    playerid                  awardid  yearid lgid namefirst namelast  \
0  leylaji99  TSN Manager of the Year    1988   NL       Jim  Leyland   
1  leylaji99  TSN Manager of the Year    1990   NL       Jim  Leyland   
2  leylaji99  TSN Manager of the Year    1992   NL       Jim  Leyland   
3  leylaji99  TSN Manager of the Year    2006   AL       Jim  Leyland   
4  johnsda02  TSN Manager of the Year    1997   AL     Davey  Johnson   
5  johnsda02  TSN Manager of the Year    2012   NL     Davey  Johnson   

                   name  
0    Pittsburgh Pirates  
1    Pittsburgh Pirates  
2    Pittsburgh Pirates  
3        Detroit Tigers  
4     Baltimore Orioles  
5  Washington Nationals  


# Bonus

## Problem 5: Colleges in Tennessee

### Part (a)

### Part (b)

## Problem 6: Correlating Wins and Salary

### Part (a)

### Part (b)

### Part (c)

## Problem 7: Lefty Pitchers

### Part (a)

### Part (b)

### Part (c)